lib_eio/sync.ml

```diff
@@ -412,6 +447,37 @@ let take (t : _ t) =
     take_suspend t (Long (Q.next_suspend t.consumers))
   )
 
+let select_of_many (type a b) (ts: (a t * (a -> b)) list) =
+  let finished = Atomic.make false in
+  let cancel_fns = ref [] in
+  let add_cancel_fn fn = cancel_fns := (fn :: !cancel_fns) in
+  let cancel_all () = List.iter (fun fn -> fn ()) !cancel_fns in
+  let wait ctx enqueue ((t, f): (a t * (a -> b))) = begin
+    if (Atomic.fetch_and_add t.balance (-1)) > 0 then (
+      (* have item, can cancel remaining stream waiters*)
+      if Atomic.compare_and_set finished false true then (
+        cancel_all ();
+        (* Item is available, take it over from producer. *)
+        let cell = Q.next_resume t.producers in
+        (* [in_transition] is [None] so we spin-wait on the item being available.
+           This is a rare scenario. *)
+        let v = consumer_resume_cell t cell
+            ~success:(fun it -> it.kp (Ok true); it.v)
+            ?in_transition:None in
+        enqueue (Ok (f v))
+      ) else (
+        (* restore old balance, because another stream was ready first. *)
+        ignore (Atomic.fetch_and_add t.balance (+1))
```

Comment:
This isn't safe. Some other domain may have seen the updated value and be relying on it.